### ISSUE TITLE
Stemming only works on allText in content layer

### ISSUE THREAD

**Issue by ComLock** [enonic/xp#8876]
I have tried https://github.com/ComLock/app-stemming-example on XP 7.6.1 and stemming doesn't work.

Current docs I can find:
https://developer.enonic.com/docs/xp/stable/storage/indexing#stemmed
https://developer.enonic.com/docs/xp/stable/storage/noql#stemmed

Future doc supposed to work:
https://developer.enonic.com/docs/xp/next/storage/indexing#languages

https://github.com/ComLock/app-stemming-example/blob/master/src/main/resources/main.es#L13-L18
fulltext: true, // Needed for stemming?
includeInAllText: true, // Needed for stemming?
languages: ['no']
stemmed: true // Not reflected in node, nor in documentation, so this some core dev must have told me?

https://github.com/ComLock/app-stemming-example/blob/master/src/main/resources/main.es#L114
query: `stemmed('_allText', '${word}', 'OR', 'no')`

When using data toolbox I can see there is no 
_alltext._stemmed_no or any other fieldName._stemmed_no
under Display Search Index Document

Also when exporting and inspecting node.xml the node.xml doesn't look anything like the _indexConfig.
So export format is wastly different from node JSON. Why???

When making a normal content site and setting language to "no" there is a _alltext._stemmed_no
But when looking at the node JSON there are no languages set under _indexConfig!

This is what I think makes stemming work for content in the exported node.xml

```
<allTextIndexConfig>
    <languages>
        <language>no</language>
    </languages>
</allTextIndexConfig>
```

**Comment by vbradnitski** [comment]
- There is a string in the new doc: 
" While setting the language for the content will only index the _allText field, setting the languages ​​in the node config will create stemmed indices for all mapped properties. See node create function. "
Which means we don't create stemmed `_allText` index for nodes, only for contents. So `stemmed('_allText')` function will return nothing for node.

- The is a lie in the new doc: custom index cannot be created for the `_name` field, it's always fulltext. So it can't be stemmed. Doc will be fixed.

- I don't know why XML export format diffs from JSON lib format, looks like XML was migrated from elder Enonic XP versions and JSON was introduced later. Not all things which good for JSON are always suitable for XML.

**Comment by rymsha** [comment]
Documentation issue created https://github.com/enonic/doc-xp/issues/307

**Comment by ComLock** [comment]
@sigdestad Can you have a look at this. If I understand correctly nothing has changed, stemming is still impossible for the node layer?

**Comment by sigdestad** [comment]
@rymsha something is obviously not working as expected. Could we arrange a meeting to clarify things related to stemming for med and CWE?

**Comment by alansemenov** [comment]
@ComLock from what I see [here](https://github.com/ComLock/app-stemming-example/blob/master/src/main/resources/main.es#L93), you are still trying to stem the _node field. This will not work, as Slava described in the comment above (and this is what he fixed in the docs too).

If you have an example where stemming doesn't work for a field that is **supposed** to work, please commit this to your app's repo and we will look at the code.

**Comment by ComLock** [comment]
Are there automatic regression tests for stemming on the node layer somewhere, it would be nice to look at some working example code. (since my example code has flaws in it)

**Comment by ComLock** [comment]
> custom index cannot be created for the _name field, it's always fulltext

What does that mean? You can't both have a fulltext and stemmed index of a field? So I have to set fulltext to false in indexConfig?

**Comment by ComLock** [comment]
Got it working, looking into why. Maybe connection.refresh();

**Comment by ComLock** [comment]
> So stemmed('_allText') function will return nothing for node.

@sigdestad If I understand correctly even though I say includeInAllText: true on some field, there will be no _alltext._stemmed_no. I can live with that in explorer. But might be something we want in the future?

**Comment by sigdestad** [comment]
So, includeInAllText just indicates if a property should be included when creating the _allText "virtual field". What we need is that _allText get's stemmed.

**Comment by ComLock** [comment]
Something like: If any field indexConfig (including default) both have includeInAllText: true and any lang in languages: [] then there needs to be created an _alltext._stemmed_LANG per LANG

**Comment by vbradnitski** [comment]
`includeInAllText` and `languages` are separate configs: `includeInAllText` indicates if field value will be add to `_allText` index and `languages` sets array of stemmed indices for only this particular field. Adding stemmed index for special  `_allText` field was designed as a content level thing only for single `language` content field value.

**Comment by sigdestad** [comment]
But doesn't content API use node API directly? How can it do something else than what node api supports?

**Comment by sigdestad** [comment]
Basically, what I would like to know is how to specify that -_allText should be stemmed?

**Comment by vbradnitski** [comment]
yes, content API uses node API, but _allText stemming feature is done in the inner core-content module, and this functionality wasn't been open neither in content or node js lib. It was decided to make it a content-specific feature. So there is no way to influence `_allText` indexing directly for now.

**Comment by sigdestad** [comment]
Ok, so we basically need to define a proper solution for this in the node API, implement and document it.

There is at least one known "problem" with this in the export/import xml that we also need to look into

**Comment by rymsha** [comment]
It is still not exactly clear what to do...
- proper solution for "this", but what is "this"? Do we want stemming on other fields? 
- what is exactly wrong with import/export? does it not preserve stemming configuration?

**Comment by ComLock** [comment]
I don't care that much about the import/export syntax.

But since Explorer 4 uses _alltext as the default search, it would be nice if stemming of _alltext on the node layer worked from the get go without people having to boost/query specific stemmed fields in order for stemming to work...

So I think the label of Documentation is misleading. For me this is rather a Feature Request or similar.

**Comment by rymsha** [comment]
export looks like this
```
<node>
    <id>f4ff7f86-e19d-41de-bc0e-ec5ed42c90b3</id>
    <childOrder>_ts DESC</childOrder>
    <nodeType>default</nodeType>
    <timestamp>2025-12-05T13:46:14.637Z</timestamp>
    <permissions>
        ...
    </permissions>
    <data>
        <string name="title">test node title</string>
    </data>
    <indexConfigs>
        <defaultConfig>
            <decideByType>true</decideByType>
            <enabled>true</enabled>
            <nGram>false</nGram>
            <fulltext>false</fulltext>
            <includeInAllText>true</includeInAllText>
            <languages>
                <language>en</language>
            </languages>
        </defaultConfig>
        <pathIndexConfigs/>
        <allTextIndexConfig>
            <languages>
                <language>ru</language>
            </languages>
        </allTextIndexConfig>
    </indexConfigs>
</node>
```
Notice `indexConfigs.defaultConfig.languages.language = en`
`title` field will have an index on `title._stemmed_en` and it will work with stemmed search.
`indexConfigs.allTextIndexConfig.languages.language = ru` configures _allText field stemming

How language is set in export is not obscure - but truly undocumented as well as the entire XML export format.

So, everything works including _allText config via Node API (at least in Java)

Documentation also exisis
https://developer.enonic.com/docs/xp/stable/storage/indexing#languages

@ComLock was struggling with it because he forgot to call refresh https://github.com/enonic/xp/issues/8876#issuecomment-859592620 as node API is eventually consistent by default - for performance reasons,


The only thing we can improve now - is to cross-link documentation between how to create document with stemming and how to use stemmed search.
